src/Symfony/Component/Messenger/Middleware/LockMiddleware.php

```diff
@@ -0,0 +1,90 @@
+use Symfony\Component\Messenger\Message\TTLAwareLockableMessageInterface;
+use Symfony\Component\Messenger\Stamp\LockStamp;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+
+final class LockMiddleware implements MiddlewareInterface
+{
+    private ?LockFactory $lockFactory = null;
+
+    public function __construct(
+        ?LockFactory $lockFactory,
+    ) {
+        $this->lockFactory = $lockFactory;
+    }
+
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        if (null === $this->lockFactory) {
+            return $stack->next()->handle($envelope, $stack);
+        }
+
+        $message = $envelope->getMessage();
+
+        if (null === $envelope->last(ReceivedStamp::class)) {
+            if ($message instanceof LockableMessageInterface) {
+                // If we're trying to dispatch a lockable message.
+                $keyResource = $message->getKey();
+
+                if (null !== $keyResource) {
+                    $key = new Key($keyResource);
+
+                    // The acquire call must be done before stamping the message
+                    // in order to have the full state of the key in the stamp.
+                    $lock = $message instanceof TTLAwareLockableMessageInterface
+                        ? $this->lockFactory->createLockFromKey($key, $message->getTTL(), autoRelease: false)
+                        : $this->lockFactory->createLockFromKey($key, autoRelease: false);
+
+                    if (!$lock->acquire()) {
+                        return $envelope;
+                    }
+
+                    // The acquire call must be done before stamping the message
+                    // in order to have the full state of the key in the stamp.
+                    $envelope = $envelope->with(new LockStamp($key, $message->shouldBeReleasedBeforeHandlerCall()));
+                }
+            }
+        } else {
+            $this->releaseLock($envelope, true);
+        }
+
+        try {
+            $envelope = $stack->next()->handle($envelope, $stack);
+        } finally {
+            // If we've received a lockable message, we're releasing it.
+            if (null !== $envelope->last(ReceivedStamp::class)) {
+                $this->releaseLock($envelope, false);
+            }
+        }
+
+        return $envelope;
+    }
+
+    private function releaseLock(Envelope $envelope, bool $beforeHandlerCall): void
+    {
+        $stamp = $envelope->last(LockStamp::class);
+        if ($stamp instanceof LockStamp && $stamp->shouldBeReleasedBeforHandlerCall() === $beforeHandlerCall) {
+            $message = $envelope->getMessage();
+            $lock = $message instanceof TTLAwareLockableMessageInterface
+                ? $this->lockFactory->createLockFromKey($stamp->getKey(), $message->getTTL(), autoRelease: false)
+                : $this->lockFactory->createLockFromKey($stamp->getKey(), autoRelease: false);
+            $lock->release();
```

Comment:
we want to release the lock, no need to set `autoRelease: false` nor setting the TTL.

```suggestion
            $lock = $this->lockFactory->createLockFromKey($stamp->getKey())
            $lock->release();
```